src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs

```diff
@@ -112,189 +72,211 @@ public static unsafe bool ListChildPids(int ppid, out int[] buffer)
-                {
-                    if (processEntry.ParentProcessID == ppid)
-                    {
-                        try
-                        {
-                            Process c = Process.GetProcessById(processEntry.ProcessID);
-                            if (c.ProcessName.Equals(childName, StringComparison.OrdinalIgnoreCase))
-                            {
-                                child = c;
-                                return true;
-                            }
-                            c.Dispose();
-                        }
-                        catch {}
-                    }
+        public const string COLLECT_DUMPS_ENVIRONMENT_VAR = "__CollectDumps";
+        public const string CRASH_DUMP_FOLDER_ENVIRONMENT_VAR = "__CrashDumpFolder";
 
-                    success = Kernel32.Process32NextW(snapshot, ref processEntry);
-                }
+        static bool CollectCrashDump(Process process, string path)
+        {
+            ProcessStartInfo createdumpInfo = null;
+            string coreRoot = Environment.GetEnvironmentVariable("CORE_ROOT");
+            string createdumpPath = Path.Combine(coreRoot, "createdump");
+            string arguments = $"--name \"{path}\" {process.Id} --withheap";
+            Process createdump = new Process();
 
-                child = null;
-                return false;
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                createdump.StartInfo.FileName = createdumpPath + ".exe";
+                createdump.StartInfo.Arguments = arguments;
             }
-            finally
+            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux) || RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
             {
-                Kernel32.CloseHandle(snapshot);
+                createdump.StartInfo.FileName = "sudo";
+                createdump.StartInfo.Arguments = $"{createdumpPath} " + arguments;
             }
-        }
 
-        static bool TryFindChildProcessByNameLinux(Process process, string childName, out Process child)
-        {
-            Queue<string> childrenFilesToCheck = new Queue<string>();
+            createdump.StartInfo.UseShellExecute = false;
+            createdump.StartInfo.RedirectStandardOutput = true;
+            createdump.StartInfo.RedirectStandardError = true;
 
-            childrenFilesToCheck.Enqueue($"/proc/{process.Id}/task/{process.Id}/children");
+            Console.WriteLine($"Invoking: {createdump.StartInfo.FileName} {createdump.StartInfo.Arguments}");
+            createdump.Start();
 
-            while (childrenFilesToCheck.Count != 0)
-            {
-                string childrenFile = childrenFilesToCheck.Dequeue();
+            Task<string> copyOutput = createdump.StandardOutput.ReadToEndAsync();
+            Task<string> copyError = createdump.StandardError.ReadToEndAsync();
+            bool fSuccess = createdump.WaitForExit(DEFAULT_TIMEOUT * 2);
```

Comment:
20 min to dump a process sounds a little excessive. Even if it took that long to take a dump, uploading and downloading/loading it would be rather impractical. `WaitForExit` also blocks the thread which if we have a bad deadlock we'll see hangs in CI with lots of threads creating dumps. I haven't tried createdump on macOS to give you a good estimate of a good timeout here (lldb is pretty awful..., don't know how it compares).

Reply:
Ah, the `* 2` was only added when I was testing the functionality and trying to pinpoint some timing issues.  I doubt it would take 20 minutes in _most_ cases for coreclr pri0 tests.  The original timeout was `DEFAULT_TIMEOUT` (10 minutes), which still seems extreme.  I'll put it back to 10 minutes, but we might want to revisit whether that still makes sense.

Reply:
This is still a bit scary to me for dumping timeouts.

Reply:
> `WaitForExit` also blocks the thread which if we have a bad deadlock we'll see hangs in CI with lots of threads creating dumps.

I assume you're talking about `createdump` deadlocking here?  We could follow the timeout up with a call to `Process.Kill()` in the event of timeout to make sure we don't leave any zombied `createdump` processes in CI.  I'm guessing the the file IO is going to be the limiting factor when it comes to how long it takes for `createdump` to run to completion.

Reply:
Hmmm... Now that I'm thinking about it... I don't think timed-out tests are killed either.  Let me add that and run one more CI with the purposefully failing test.